{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6cd8836d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import qiime2 as q2\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from biom import Table\n",
    "from skbio import DistanceMatrix\n",
    "from validation_util import set_root_path, decorate_path, load_run, remove_singletons\n",
    "from qiime2.plugins.diversity.actions import core_metrics_phylogenetic, alpha_rarefaction\n",
    "from qiime2.plugins.diversity.visualizers import beta_group_significance, adonis\n",
    "from bp import parse_newick, write_newick\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c8cba16b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# biomass can be either [high, low]\n",
    "BP='150-bp'\n",
    "BIOMASS='low'\n",
    "DESCRIPTION=None\n",
    "runs = [1,2,3]\n",
    "# labs = ['kl']\n",
    "labs = ['kl', 'core']\n",
    "ROOT_PATH=set_root_path(BP, [BIOMASS, DESCRIPTION], runs, labs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bf8bbd84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8568, 165) (4851, 165) 5016.666666666667\n",
      "(15454, 175) (5947, 175) 19526.72\n",
      "(12411, 172) (7219, 172) 8440.53488372093\n",
      "(12484, 175) (5025, 175) 28245.257142857143\n",
      "(9497, 172) (5685, 172) 6021.0058139534885\n",
      "(19918, 175) (8007, 175) 25308.245714285713\n",
      "(24044, 1034)\n",
      "(19205, 1034)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>col</th>\n",
       "      <th>row</th>\n",
       "      <th>sex</th>\n",
       "      <th>diet</th>\n",
       "      <th>round</th>\n",
       "      <th>title</th>\n",
       "      <th>water</th>\n",
       "      <th>empo_1</th>\n",
       "      <th>empo_2</th>\n",
       "      <th>empo_3</th>\n",
       "      <th>...</th>\n",
       "      <th>physical_specimen_remaining</th>\n",
       "      <th>genomic_spike_in_copy_number</th>\n",
       "      <th>genomic_spike_in_description</th>\n",
       "      <th>metagenomic_spike_in_replicate</th>\n",
       "      <th>metagenomic_spike_in_copy_number</th>\n",
       "      <th>metagenomic_spike_in_amount_ng_per_ul</th>\n",
       "      <th>qiita_study_id</th>\n",
       "      <th>original_id</th>\n",
       "      <th>run</th>\n",
       "      <th>lab</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>run-1-kl14332.361162779</th>\n",
       "      <td>11</td>\n",
       "      <td>C</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>1_2</td>\n",
       "      <td>Matrix_pipeline_validation</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Free-living</td>\n",
       "      <td>Non-saline</td>\n",
       "      <td>Surface (non-saline)</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>syndna plasmids</td>\n",
       "      <td>1</td>\n",
       "      <td>1.61e+07</td>\n",
       "      <td>0.02</td>\n",
       "      <td>14332</td>\n",
       "      <td>14332.361162779</td>\n",
       "      <td>r1</td>\n",
       "      <td>kl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run-1-kl14332.361162780</th>\n",
       "      <td>11</td>\n",
       "      <td>A</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>1_2</td>\n",
       "      <td>Matrix_pipeline_validation</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Free-living</td>\n",
       "      <td>Non-saline</td>\n",
       "      <td>Surface (non-saline)</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>2.76e+06</td>\n",
       "      <td>syndna plasmids</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>14332</td>\n",
       "      <td>14332.361162780</td>\n",
       "      <td>r1</td>\n",
       "      <td>kl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run-1-kl14332.361162792</th>\n",
       "      <td>8</td>\n",
       "      <td>G</td>\n",
       "      <td>female</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>1_2</td>\n",
       "      <td>Matrix_pipeline_validation</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Animal</td>\n",
       "      <td>Animal surface</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>14332</td>\n",
       "      <td>14332.361162792</td>\n",
       "      <td>r1</td>\n",
       "      <td>kl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run-1-kl14332.361162794</th>\n",
       "      <td>12</td>\n",
       "      <td>O</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>1_2</td>\n",
       "      <td>Matrix_pipeline_validation</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Free-living</td>\n",
       "      <td>Non-saline</td>\n",
       "      <td>Surface (non-saline)</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>syndna plasmids</td>\n",
       "      <td>2</td>\n",
       "      <td>1.61e+07</td>\n",
       "      <td>0.02</td>\n",
       "      <td>14332</td>\n",
       "      <td>14332.361162794</td>\n",
       "      <td>r1</td>\n",
       "      <td>kl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run-1-kl14332.361162797</th>\n",
       "      <td>9</td>\n",
       "      <td>M</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>1_2</td>\n",
       "      <td>Matrix_pipeline_validation</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Free-living</td>\n",
       "      <td>Non-saline</td>\n",
       "      <td>Surface (non-saline)</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>syndna plasmids</td>\n",
       "      <td>1</td>\n",
       "      <td>2.59e+07</td>\n",
       "      <td>0.04</td>\n",
       "      <td>14332</td>\n",
       "      <td>14332.361162797</td>\n",
       "      <td>r1</td>\n",
       "      <td>kl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run-3-core14332.361164197</th>\n",
       "      <td>10</td>\n",
       "      <td>B</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>1_2</td>\n",
       "      <td>Matrix_pipeline_validation</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Free-living</td>\n",
       "      <td>Non-saline</td>\n",
       "      <td>Surface (non-saline)</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>5.51e+06</td>\n",
       "      <td>syndna plasmids</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>14332</td>\n",
       "      <td>14332.361164197</td>\n",
       "      <td>r3</td>\n",
       "      <td>core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run-3-core14332.363197873</th>\n",
       "      <td>6</td>\n",
       "      <td>N</td>\n",
       "      <td>female</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>1_2</td>\n",
       "      <td>Matrix_pipeline_validation</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Animal</td>\n",
       "      <td>Animal surface</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>2.21e+07</td>\n",
       "      <td>syndna plasmids</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>14332</td>\n",
       "      <td>14332.363197873</td>\n",
       "      <td>r3</td>\n",
       "      <td>core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run-3-core14332.CONTROL.2.2.2</th>\n",
       "      <td>5</td>\n",
       "      <td>C</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>1_2</td>\n",
       "      <td>Matrix_pipeline_validation</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Control</td>\n",
       "      <td>Positive</td>\n",
       "      <td>Single strain</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>14332</td>\n",
       "      <td>14332.CONTROL.2.2.2</td>\n",
       "      <td>r3</td>\n",
       "      <td>core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run-3-core14332.CONTROL.6.2.2</th>\n",
       "      <td>10</td>\n",
       "      <td>I</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>1_2</td>\n",
       "      <td>Matrix_pipeline_validation</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Control</td>\n",
       "      <td>Positive</td>\n",
       "      <td>Single strain</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>14332</td>\n",
       "      <td>14332.CONTROL.6.2.2</td>\n",
       "      <td>r3</td>\n",
       "      <td>core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run-3-core14332.361162766</th>\n",
       "      <td>4</td>\n",
       "      <td>G</td>\n",
       "      <td>male</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>1_2</td>\n",
       "      <td>Matrix_pipeline_validation</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Animal</td>\n",
       "      <td>Animal surface</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>4.41e+07</td>\n",
       "      <td>syndna plasmids</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>14332</td>\n",
       "      <td>14332.361162766</td>\n",
       "      <td>r3</td>\n",
       "      <td>core</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>990 rows Ã— 77 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               col row             sex            diet round  \\\n",
       "sample_name                                                                    \n",
       "run-1-kl14332.361162779         11   C  not applicable  not applicable   1_2   \n",
       "run-1-kl14332.361162780         11   A  not applicable  not applicable   1_2   \n",
       "run-1-kl14332.361162792          8   G          female  not applicable   1_2   \n",
       "run-1-kl14332.361162794         12   O  not applicable  not applicable   1_2   \n",
       "run-1-kl14332.361162797          9   M  not applicable  not applicable   1_2   \n",
       "...                            ...  ..             ...             ...   ...   \n",
       "run-3-core14332.361164197       10   B  not applicable  not applicable   1_2   \n",
       "run-3-core14332.363197873        6   N          female  not applicable   1_2   \n",
       "run-3-core14332.CONTROL.2.2.2    5   C  not applicable  not applicable   1_2   \n",
       "run-3-core14332.CONTROL.6.2.2   10   I  not applicable  not applicable   1_2   \n",
       "run-3-core14332.361162766        4   G            male  not applicable   1_2   \n",
       "\n",
       "                                                    title           water  \\\n",
       "sample_name                                                                 \n",
       "run-1-kl14332.361162779        Matrix_pipeline_validation  not applicable   \n",
       "run-1-kl14332.361162780        Matrix_pipeline_validation  not applicable   \n",
       "run-1-kl14332.361162792        Matrix_pipeline_validation  not applicable   \n",
       "run-1-kl14332.361162794        Matrix_pipeline_validation  not applicable   \n",
       "run-1-kl14332.361162797        Matrix_pipeline_validation  not applicable   \n",
       "...                                                   ...             ...   \n",
       "run-3-core14332.361164197      Matrix_pipeline_validation  not applicable   \n",
       "run-3-core14332.363197873      Matrix_pipeline_validation  not applicable   \n",
       "run-3-core14332.CONTROL.2.2.2  Matrix_pipeline_validation  not applicable   \n",
       "run-3-core14332.CONTROL.6.2.2  Matrix_pipeline_validation  not applicable   \n",
       "run-3-core14332.361162766      Matrix_pipeline_validation  not applicable   \n",
       "\n",
       "                                        empo_1      empo_2  \\\n",
       "sample_name                                                  \n",
       "run-1-kl14332.361162779            Free-living  Non-saline   \n",
       "run-1-kl14332.361162780            Free-living  Non-saline   \n",
       "run-1-kl14332.361162792        Host-associated      Animal   \n",
       "run-1-kl14332.361162794            Free-living  Non-saline   \n",
       "run-1-kl14332.361162797            Free-living  Non-saline   \n",
       "...                                        ...         ...   \n",
       "run-3-core14332.361164197          Free-living  Non-saline   \n",
       "run-3-core14332.363197873      Host-associated      Animal   \n",
       "run-3-core14332.CONTROL.2.2.2          Control    Positive   \n",
       "run-3-core14332.CONTROL.6.2.2          Control    Positive   \n",
       "run-3-core14332.361162766      Host-associated      Animal   \n",
       "\n",
       "                                             empo_3  ...  \\\n",
       "sample_name                                          ...   \n",
       "run-1-kl14332.361162779        Surface (non-saline)  ...   \n",
       "run-1-kl14332.361162780        Surface (non-saline)  ...   \n",
       "run-1-kl14332.361162792              Animal surface  ...   \n",
       "run-1-kl14332.361162794        Surface (non-saline)  ...   \n",
       "run-1-kl14332.361162797        Surface (non-saline)  ...   \n",
       "...                                             ...  ...   \n",
       "run-3-core14332.361164197      Surface (non-saline)  ...   \n",
       "run-3-core14332.363197873            Animal surface  ...   \n",
       "run-3-core14332.CONTROL.2.2.2         Single strain  ...   \n",
       "run-3-core14332.CONTROL.6.2.2         Single strain  ...   \n",
       "run-3-core14332.361162766            Animal surface  ...   \n",
       "\n",
       "                              physical_specimen_remaining  \\\n",
       "sample_name                                                 \n",
       "run-1-kl14332.361162779                             False   \n",
       "run-1-kl14332.361162780                             False   \n",
       "run-1-kl14332.361162792                             False   \n",
       "run-1-kl14332.361162794                             False   \n",
       "run-1-kl14332.361162797                             False   \n",
       "...                                                   ...   \n",
       "run-3-core14332.361164197                           False   \n",
       "run-3-core14332.363197873                           False   \n",
       "run-3-core14332.CONTROL.2.2.2                        True   \n",
       "run-3-core14332.CONTROL.6.2.2                        True   \n",
       "run-3-core14332.361162766                           False   \n",
       "\n",
       "                              genomic_spike_in_copy_number  \\\n",
       "sample_name                                                  \n",
       "run-1-kl14332.361162779                     not applicable   \n",
       "run-1-kl14332.361162780                           2.76e+06   \n",
       "run-1-kl14332.361162792                     not applicable   \n",
       "run-1-kl14332.361162794                     not applicable   \n",
       "run-1-kl14332.361162797                     not applicable   \n",
       "...                                                    ...   \n",
       "run-3-core14332.361164197                         5.51e+06   \n",
       "run-3-core14332.363197873                         2.21e+07   \n",
       "run-3-core14332.CONTROL.2.2.2               not applicable   \n",
       "run-3-core14332.CONTROL.6.2.2               not applicable   \n",
       "run-3-core14332.361162766                         4.41e+07   \n",
       "\n",
       "                               genomic_spike_in_description  \\\n",
       "sample_name                                                   \n",
       "run-1-kl14332.361162779                     syndna plasmids   \n",
       "run-1-kl14332.361162780                     syndna plasmids   \n",
       "run-1-kl14332.361162792                      not applicable   \n",
       "run-1-kl14332.361162794                     syndna plasmids   \n",
       "run-1-kl14332.361162797                     syndna plasmids   \n",
       "...                                                     ...   \n",
       "run-3-core14332.361164197                   syndna plasmids   \n",
       "run-3-core14332.363197873                   syndna plasmids   \n",
       "run-3-core14332.CONTROL.2.2.2                not applicable   \n",
       "run-3-core14332.CONTROL.6.2.2                not applicable   \n",
       "run-3-core14332.361162766                   syndna plasmids   \n",
       "\n",
       "                              metagenomic_spike_in_replicate  \\\n",
       "sample_name                                                    \n",
       "run-1-kl14332.361162779                                    1   \n",
       "run-1-kl14332.361162780                       not applicable   \n",
       "run-1-kl14332.361162792                       not applicable   \n",
       "run-1-kl14332.361162794                                    2   \n",
       "run-1-kl14332.361162797                                    1   \n",
       "...                                                      ...   \n",
       "run-3-core14332.361164197                     not applicable   \n",
       "run-3-core14332.363197873                     not applicable   \n",
       "run-3-core14332.CONTROL.2.2.2                 not applicable   \n",
       "run-3-core14332.CONTROL.6.2.2                 not applicable   \n",
       "run-3-core14332.361162766                     not applicable   \n",
       "\n",
       "                               metagenomic_spike_in_copy_number  \\\n",
       "sample_name                                                       \n",
       "run-1-kl14332.361162779                                1.61e+07   \n",
       "run-1-kl14332.361162780                          not applicable   \n",
       "run-1-kl14332.361162792                          not applicable   \n",
       "run-1-kl14332.361162794                                1.61e+07   \n",
       "run-1-kl14332.361162797                                2.59e+07   \n",
       "...                                                         ...   \n",
       "run-3-core14332.361164197                        not applicable   \n",
       "run-3-core14332.363197873                        not applicable   \n",
       "run-3-core14332.CONTROL.2.2.2                    not applicable   \n",
       "run-3-core14332.CONTROL.6.2.2                    not applicable   \n",
       "run-3-core14332.361162766                        not applicable   \n",
       "\n",
       "                              metagenomic_spike_in_amount_ng_per_ul  \\\n",
       "sample_name                                                           \n",
       "run-1-kl14332.361162779                                        0.02   \n",
       "run-1-kl14332.361162780                              not applicable   \n",
       "run-1-kl14332.361162792                              not applicable   \n",
       "run-1-kl14332.361162794                                        0.02   \n",
       "run-1-kl14332.361162797                                        0.04   \n",
       "...                                                             ...   \n",
       "run-3-core14332.361164197                            not applicable   \n",
       "run-3-core14332.363197873                            not applicable   \n",
       "run-3-core14332.CONTROL.2.2.2                        not applicable   \n",
       "run-3-core14332.CONTROL.6.2.2                        not applicable   \n",
       "run-3-core14332.361162766                            not applicable   \n",
       "\n",
       "                              qiita_study_id          original_id run   lab  \n",
       "sample_name                                                                  \n",
       "run-1-kl14332.361162779                14332      14332.361162779  r1    kl  \n",
       "run-1-kl14332.361162780                14332      14332.361162780  r1    kl  \n",
       "run-1-kl14332.361162792                14332      14332.361162792  r1    kl  \n",
       "run-1-kl14332.361162794                14332      14332.361162794  r1    kl  \n",
       "run-1-kl14332.361162797                14332      14332.361162797  r1    kl  \n",
       "...                                      ...                  ...  ..   ...  \n",
       "run-3-core14332.361164197              14332      14332.361164197  r3  core  \n",
       "run-3-core14332.363197873              14332      14332.363197873  r3  core  \n",
       "run-3-core14332.CONTROL.2.2.2          14332  14332.CONTROL.2.2.2  r3  core  \n",
       "run-3-core14332.CONTROL.6.2.2          14332  14332.CONTROL.6.2.2  r3  core  \n",
       "run-3-core14332.361162766              14332      14332.361162766  r3  core  \n",
       "\n",
       "[990 rows x 77 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tables = []\n",
    "metadatas = []\n",
    "for run_i in range(len(runs)):\n",
    "    for lab_j in range(len(labs)):\n",
    "        run, lab = runs[run_i], labs[lab_j]\n",
    "        table, metadata = load_run(BP, run, lab, BIOMASS, DESCRIPTION)\n",
    "        tables.append(table)\n",
    "        metadatas.append(metadata)\n",
    "        if run_i == 0 and lab_j == 0:\n",
    "            ids = metadata['original_id']\n",
    "full_table, full_metadata= tables[0], metadatas[0]\n",
    "full_table = full_table.merge(tables[1:])\n",
    "print(full_table.shape)\n",
    "full_table = remove_singletons(full_table)\n",
    "full_table = full_table.remove_empty(axis='observation', inplace=False)\n",
    "print(full_table.shape)\n",
    "full_metadata = pd.concat(metadatas)\n",
    "metadata_f_name = decorate_path('metadata.tsv', ROOT_PATH)\n",
    "\n",
    "full_metadata.to_csv(metadata_f_name, sep='\\t', index=True)\n",
    "full_metadata.loc[full_metadata['original_id'].isin(ids)]\n",
    "# full_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4899d0c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mImported 150-bp/low-runs-1-2-3-labs-kl-core/tree.tre as NewickDirectoryFormat to 150-bp/low-runs-1-2-3-labs-kl-core/tree.qza\u001b[0m\r\n"
     ]
    }
   ],
   "source": [
    "tree_path=os.path.join(BP, '{BP}.tre'.format(BP=BP))\n",
    "tree = parse_newick(open(tree_path).read())\n",
    "tree.shear(set(full_table.ids(axis='observation')))\n",
    "tree_f_name = decorate_path('tree.tre', ROOT_PATH)\n",
    "q2_tree_f_name = decorate_path('tree.qza', ROOT_PATH)\n",
    "\n",
    "with open(tree_f_name, 'w') as f:\n",
    "    write_newick(tree, f, include_edge=True)\n",
    "    \n",
    "! qiime tools import \\\n",
    "--type 'Phylogeny[Rooted]' \\\n",
    "--input-path {tree_f_name} \\\n",
    "--output-path {q2_tree_f_name} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8a27c260",
   "metadata": {},
   "outputs": [],
   "source": [
    "q2_full_table = q2.Artifact.import_data(type='FeatureTable[Frequency]', view=full_table)\n",
    "q2_full_metadata = q2.Metadata.load(metadata_f_name)\n",
    "q2_full_tree = q2.Artifact.load(q2_tree_f_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0ac0dd70",
   "metadata": {},
   "outputs": [],
   "source": [
    "rf_depths = [1500]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4768b2c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#core diversity metrics @ different sequencing depths\n",
    "for rf_depth in rf_depths:\n",
    "    core_results = core_metrics_phylogenetic(q2_full_table,\n",
    "                                             q2_full_tree,\n",
    "                                             rf_depth,\n",
    "                                             q2_full_metadata)\n",
    "    \n",
    "    results_dir_path = decorate_path(\n",
    "        'core-metrics-%i'%rf_depth, ROOT_PATH\n",
    "    )\n",
    "    if os.path.isdir(results_dir_path):\n",
    "        shutil.rmtree(results_dir_path)\n",
    "    os.mkdir(results_dir_path)\n",
    "    \n",
    "    for name, q2art in zip(core_results._fields, core_results):\n",
    "        out_ = os.path.join(results_dir_path,name)\n",
    "        q2art.save(out_)\n",
    "    rare_table_path = '{results_path}/rarefied_table.qza'.format(results_path=results_dir_path)\n",
    "    table = q2.Artifact.load(rare_table_path).view(Table)\n",
    "    print(table.shape)\n",
    "    table = table.remove_empty(axis='observation', inplace=False)\n",
    "    print(table.shape)\n",
    "    q2_table = q2.Artifact.import_data(type='FeatureTable[Frequency]', view=table)\n",
    "    q2_table.save(rare_table_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7368add",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compare between labs\n",
    "factors=['storage_solution','extraction_protocol']\n",
    "major_factors=['sample_type', 'host_subject_id']\n",
    "states=['lab', 'run']\n",
    "# states=['lab']\n",
    "\n",
    "lab_formula='\"{major} + {factors} + {lab} +({i_major} + {i_factors}):{i_lab} \"'.format(\n",
    "    major='*'.join(major_factors),\n",
    "    i_major='+'.join(major_factors),\n",
    "    lab='*'.join(states),\n",
    "    i_lab='*'.join(states),\n",
    "    factors='*'.join(factors),\n",
    "    i_factors='+'.join(factors)\n",
    ")\n",
    "#     distance_metrics = ['bray_curtis', 'jaccard', 'unweighted_unifrac', 'weighted_unifrac']\n",
    "for rf_depth in rf_depths:\n",
    "    results_dir_path = decorate_path(\n",
    "        'core-metrics-%i'%rf_depth, ROOT_PATH\n",
    "    )\n",
    "        \n",
    "#     distance_metrics = ['bray_curtis', 'jaccard', 'unweighted_unifrac', 'weighted_unifrac']\n",
    "    distance_metrics = ['unweighted_unifrac']\n",
    "    for d_metric in distance_metrics:\n",
    "        d_matrix = os.path.join(\n",
    "            results_dir_path, \n",
    "            '{metric}_distance_matrix.qza'.format(metric=d_metric))\n",
    "        !qiime diversity adonis \\\n",
    "        --i-distance-matrix {d_matrix} \\\n",
    "        --m-metadata-file {metadata_f_name} \\\n",
    "        --p-formula {lab_formula} \\\n",
    "        --o-visualization {results_dir_path}/adonis-run-{d_metric}.qzv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f282b9b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
